devito/arguments.py

```diff
@@ -75,78 +79,23 @@ class ScalarArgument(Argument):
 
     is_ScalarArgument = True
 
-    def __init__(self, name, provider, reducer=lambda old, new: new, default_value=None):
-        super(ScalarArgument, self).__init__(name, provider, default_value)
-        self.reducer = reducer
-
-    def reset(self):
-        super(ScalarArgument, self).reset()
-        self._frozen = False
-
-    def verify(self, value, enforce=False):
-        """ ScalarArgument.verify is called from many different places, with different
-            values. e.g.
-            1. From inside Operator.arguments(), with an accompanying user-provided
-            value from kwargs
-            2. From the verify of a Dimension object, with an accompanying "derived"
-            value, which is the best-guess for this ScalarArgument's value that
-            the Dimension can come up with.
-            In the above two cases, the value passed in case 2 is "optional", since
-            a dimension "belongs" to multiple Function objects and each object might
-            lead to a different best-guess. During normal behaviour, verify is expected
-            to be called multiple times with different values and this method will
-            combine all the values received in the different invocations in a configurable
-            (self.reducer) way. However, if the user has explicitly passed in a value
-            (case 1 above), that value has to be respected and this best-guess behaviour
-            described above needs to be switched off.
-            Another thing to note is that, once a ScalarArgument has been passed a value
-            with enforce=True, any further calls to verify should not change the stored
-            value. The self._frozen flag is used to ensure that.
-            :param enforce: Turn off the best-guess behaviour and just use the value
-                            passed, ignoring any previously stored values.
-        """
-        if value is not None:
-            if self._value is not None and not enforce:
-                # We already have a value and the value passed
-                # here is not enforced (i.e. optional)
-                # If we aren't frozen, use the reducer
-                if not self._frozen:
-                    self._value = self.reducer(self._value, value)
-            else:
-                # Either this is the first time we're getting a value (self.value = None)
-                # or this is an enforced value
-                if enforce or not self._frozen:
-                    # If we're frozen, ignore, else use this value directly
-                    self._value = value
-                    if enforce:
-                        # If we were forced to use this value, make sure we don't change
-                        # it in the same invocation
-                        self._frozen = True
-        return self._value is not None
+    def __init__(self, name, dependencies, dtype=np.int32):
```

Comment:
for consistency, it'd help to have a (default) ``provider`` for ``ScalarArgument`` too. Then you could move some code into the superclass constructor (``Argument``) and have a property returning the ``dtype`` directly from the provider. Not sure if this is doable

Reply:
`provider`s don't make sense for `ScalarArgument`s.